### PR TITLE
Prevent script window crash

### LINKED ISSUE #22009: Python script window execution causes MantidPlot to crash
The following Python scripts crash MantidPlot when executed via the script window:
This is a cleaner re-write of issue https://github.com/mantidproject/mantid/issues/21955.

A clearly not valid Python code that should give an error is (actually always crashes MantidPlot):
```Python
A = CreateSampleWorkspace()
B = A.getRun().getProperty('run_title')
A = CreateSampleWorkspace()
A.setTitle(B)
```

A possibly valid Python code is (actually crashes MantidPlot as well - maybe not directly, but consecutive or sequential further executions):
```Python
A = CreateSampleWorkspace()
B = A.getRun().getProperty('run_title')
A = CreateSampleWorkspace()
A.setTitle(B.value)
```

Experienced on Ubuntu 16.04, master

### PR DESCRIPTION
**Pull request by vreimund**
Fixes a bug. See issue(s) for more information.

**To test:**
Make sure this does not crash MantidPlot.

```
A = CreateSampleWorkspace()
B = A.getRun().getProperty('run_title')
A = CreateSampleWorkspace()
A.setTitle(B)
```

**Try to provoke a crash using the Python script window.**

Fixes #22009 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/PythonScript.cpp [hunk 1]
```diff
@@ -658,9 +657,9 @@ bool PythonScript::executeString() {
   } else {
     emit finished(MSG_FINISHED);
     success = true;
-  }
-  if (isInteractive()) {
-    generateAutoCompleteList();
+    if (isInteractive() && !isExecuting()) {
```
Comment (martyngigg) [inline MantidPlot/src/PythonScript.cpp]: I think the `isExecuting` check here means that the code inside the `if` here will never run. I think `isExecuting` will only return false after `executeString()` has completed.

To demonstrate this first type `ws = CreateSampleWorkspace()` in to the script window and execute the code. Now on a new line start typing `ws.` and nothing happens. In the current version of Mantid the selection box of methods pops up, which is created by a call to `generateAutoCompleteList()`

**Review by martyngigg — CHANGES_REQUESTED** [review]
I'm happy that we only update the auto-complete information on successful execution of a script.

**Comment by martyngigg** [comment]
I think our main problem here is that we are hitting a real problem under the hood with how `getRun().getProperty` works. `getProperty` returns a bare pointer to a log value and when `A` is overwritten the pointer stored in `B` is no longer valid so any further accesses of this pointer will likely cause a segmentation fault.

I think your fix, without the `isExecuting` check, is still valid as it at least allows an error message to be populated.

The issue with `getProperty` is rather harder to solve and the workaround at the moment would be:
```
A = CreateSampleWorkspace()
a_title = A.getRun().getProperty('run_title').value
A = CreateSampleWorkspace()
A.setTitle(a_title)
```
so that `a_title` becomes a copy of the value in the property and having a dangling pointer in the script at all.

I suggest we try and tackle the *real* problem with `getPointer` in the next release as it is to big to put in now.

**Comment by reimundILL** [comment]
Thank you for your careful considerations, I did not check that auto-completion still works throughout.

I will follow your advice to allow auto-completion and open an issue for the remaining bug.  I hope you follow this further since I will likely not have time to do some further investigations.

Some points that stay important for me are:
- this solution does (I was really happy) not produce any crashes in no multiple ways for me.

- the script
```Python
A = CreateSampleWorkspace()
B = A.getRun().getProperty('run_title')
A = CreateSampleWorkspace()
A.setTitle(B.value)
```
in the script interpreter does not produce crashes at all (on master)
- but this script will still crash Mantid in the future when run in the script window.

The script interpreter and the script window should behave the same way, the script window should be fixed first from my point of view.

**Comment by martyngigg** [comment]
@reimundILL Thanks for re-enabling the auto completion. I agree that it's important we fix the underlying issue so that the program does not crash in any instance when accessing the log data in that way.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**